README.md
# Ahead of the Curve

Available to download on Curse by clicking [here](https://mods.curse.com/addons/wow/258441-ahead-of-the-curve) or via the [Curse client](https://www.curse.com/). You can also check out this project on [CurseForge](https://wow.curseforge.com/projects/ahead-of-the-curve).

### About

A World of Warcraft addon that automates the sending of Ahead of the Curve achievements when signing up for raid groups via the Looking for Group system.

### Features

- Automates the sending of Ahead of the Curve and Mythic Plus achievements when signing up for groups via the LFG system.
- Helps links your Mythic Plus keystone when signing up for Mythic Plus dungeon.
- Determines highest achievement found for the raid your signing up for on your account.
  - Raids: Cutting Edge, Mythic End Boss Achievement, Ahead of the Curve, Normal/LFR Achievement
  - Mythic Plus: Realm First, Keystone Master, Keystone Conqueror
- Allows you to override defaults by picking any achievement you have completed. This override will send for any Battle for Azeroth raid or Mythic Plus dungeon group you sign up for.
- When signing up for a group you have the option to send a whisper or not. If no default achievement is found and no override is set no whisper will be sent.
- Clicking on the button or icon under the Default Highest Achievements Found section will now link that achievement to your chat input box.

### Options

Found using the slash command `/aotc` or under **Interface > AddOns Tab > Ahead of the Curve > Configure Button**

### Slash Commands

- `/aotc` - Brings up the options menu

### Found a Bug?

Please open an issue on Curse: [https://wow.curseforge.com/projects/ahead-of-the-curve/issues](https://wow.curseforge.com/projects/ahead-of-the-curve/issues)


CHANGELOG.md
# Change Log

## 1.1.11 (12/28/2020)

### Improvements

- Added Shadowlands raids and dungeons.

## 1.1.10 (11/08/2020)

### Improvements

- Updated for Shadowlands pre-patch. (Thanks to gaveer!)

## 1.1.9 (08/30/2018)

### Improvements

- Fixed error that prevented users from signing up for groups.

## 1.1.8 (08/29/2018)

### Improvements

- Updated for BfA! Removed all Legion content.
- Started adding BfA raids and mythic plus dungeons.

## 1.1.7 (12/12/2017)

### Improvements

- Added missing mythic level raid detection setting in LFG (Thanks Kpi!)

## 1.1.6 (12/12/2017)

### Improvements

- Added missing Antorus achievements and id for Triunvirate (Thanks Kpi!)

## 1.1.5 (06/16/2017)

### Improvements

- Added missing Tomb of Sargeras achievements

## 1.1.4 (05/06/2017)

### Improvements

- Fixed issue that was preventing search results from displaying initially. This was due to Blizzard making some LFG functions private.

## 1.1.3 (04/05/2017)

### Improvements

- Added ids for new mythic plus dungeons (Upper & Lower Karazhan and Cathedral of Eternal Night).
- Added Tomb of Sargeras and some achievements exect for AoTC and Cutting Edge because I don't know the ids right now. Will add them once the raid actually comes out.

## 1.1.2 (03/30/2017)

### Improvements

- Updated for patch 7.2
- Added realm first achievements as highest earned for Emerald Nightmare, Trial of Valor, Nighthold, and Keystone Master.

## 1.1.1 (03/02/2017)

### Improvements

- Fixed issue with not always finding keystones in bags. This prevents the send mythic keystone option from showing up.
- Added new wording for options based upon what your signing up for. For example, signing up for mythic plus will now say "Send Mythic Plus Achievement". Previously it just said "Use Ahead of the Curve Whisper".

## 1.1.0 (03/01/2017)

### New Features

- Now you can link your mythic plus keystone from the role select dialog box if you are signing up for a mythic plus dungeon!

### Improvements

- Added mythic achievements for Xavius, Helya, and Gul'dan. So if you don't have Cutting Edge for Xavius the highest default will be the mythic kill achievement.
- Always check achievement whisper option now doesn't require a ui reload. **Breaking Change: Always check whisper dialog checkbox options will now be set back to defaults.**

## 1.0.0 (02/05/2017)

### New Features

- Added Mythic Plus and Karazhan!
- Clicking on the button or icon under Default Highest Achievements Found section will now link that achievement to your chat input box.

### Improvements

- Made search ui a bit better by showing results returned and a default option if you have never picked an override. (Thanks Velgana!)
- If no achievements are found under the Default Highest Achievements Found then a message will now appear.
- If no achievement is found under the Default Highest Achievements Found section it will no longer list that raid/dungeon.
- Minor code cleanup.
- Bliz options now has a configure button to open up the options menu.
- Added minimap icon. (Thanks Velgana!)

## 1.0.0 Beta 2 (02/02/2017)

### Improvements

- Changed default highest achievments found from input boxes to text descriptions since they can't be changed.
- Added AoTC achievment icons to default highest achievments found.
- Default highest achievments found will now update dynamically.

## 1.0.0 Beta 1 (02/01/2017)

### Features

- Automates the sending of Ahead of the Curve achievments when signing up for raid groups via the LFG system.
- Currently only works for Legion raids.
- Determines highest achievment found for the raid your signing up for on your account (Cutting Edge, Ahead of the Curve, Normal/LFR Achievement).
- Allows you to override defaults by picking any achievement you have completed. This override will send for any Legion raid you sign up for.
- When signing up for a group you have the option to send a whisper or not. If no default achievement is found and no override is set no whisper will be sent.


AheadOfTheCurve.lua
local AheadOfTheCurve = LibStub('AceAddon-3.0'):NewAddon('AheadOfTheCurve', 'AceConsole-3.0', 'AceHook-3.0', 'AceEvent-3.0')
local AceConfigRegistry = LibStub('AceConfigRegistry-3.0')
local keyStoneLink
local defaults = {
    global = {
        enable = {
            addon = true,
            override = false,
            whispers = {
                achievement = false,
                keystone = false
            }
        },
        minimap = {
            hide = true
        } 
    }
}

function AheadOfTheCurve:OnInitialize()
    self.db = LibStub('AceDB-3.0'):New('AheadOfTheCurveDB', defaults)
    
    self.achievementSearchList = {}
    local id = self.db.global.overrideAchievement
    
    if id then
        local _, name = GetAchievementInfo(id)
        self.achievementSearchList[id] = name
    else
        self.achievementSearchList[1] = 'Please search and then select an achievement from the list'
    end

    self.instances = {
        {
            ids = {691, 695, 699, 703, 705, 709, 713, 717},
            name = 'Mythic Plus (Shadowlands)',
            achievements = {14662, 14532, 14531},
            highestCompleted = nil,
        },
        {
            ids = {720, 722, 721},
            name = 'Castle Nathria',
            achievements = {14461, 14365, 14460, 14715},
            highestCompleted = nil,
        }
    }
end

function AheadOfTheCurve:OnEnable()
    self:SecureHookScript(LFGListFrame.CategorySelection.FindGroupButton, 'OnClick', 'GetLFGCategory')
    self:HookScript(LFGListFrame.SearchPanel.SignUpButton, 'OnClick', 'GetLFGInstance')
    self:HookScript(LFGListApplicationDialog.SignUpButton, 'OnClick', 'SendWhisper')
    self:RegisterEvent('ACHIEVEMENT_SEARCH_UPDATED')
    self:RegisterEvent('ACHIEVEMENT_EARNED')
    self:GetHighestDefaultAchievement()
    
    if not self.container then
        self.container = CreateFrame('Frame', 'AheadOfTheCurveDialog', LFGListApplicationDialog, BackdropTemplateMixin and "BackdropTemplate");
        self.container:SetSize(306, 50)
        self.container:SetPoint('BOTTOM', 0, -55)
        self.container:SetBackdrop({
            bgFile = 'Interface\\DialogFrame\\UI-DialogBox-Background',
            edgeFile = 'Interface\\DialogFrame\\UI-DialogBox-Border',
            tile = true, tileSize = 32, edgeSize = 32,
            insets = {left = 8, right = 8, top = 8, bottom = 8}
        })
    end
    
    if not self.checkButtonAchievement then
        self.checkButtonAchievement = CreateFrame('CheckButton', 'AheadOfTheCurveCheckBoxAchievement', self.container, 'UICheckButtonTemplate')
        self.checkButtonAchievement:SetSize(24, 24)
        self.checkButtonAchievement:SetPoint('CENTER', -85, 0)
        self.checkButtonAchievement:SetChecked(self.db.global.enable.whispers.achievement)
    end

    if not self.checkButtonKeystone then
        self.checkButtonKeystone = CreateFrame('CheckButton', 'AheadOfTheCurveCheckBoxKeystone', self.container, 'UICheckButtonTemplate')
        self.checkButtonKeystone:SetSize(24, 24)
        self.checkButtonKeystone:SetPoint('CENTER', -85, -11)
        self.checkButtonKeystone:SetChecked(self.db.global.enable.whispers.keystone)
        self.checkButtonKeystone.text:SetText('Send Mythic Plus Keystone')
        self.checkButtonKeystone.text:SetWidth(145)
    end
    
    self.container:Show()
    self.checkButtonAchievement:Show()
    self.checkButtonKeystone:Hide()
end

function AheadOfTheCurve:OnDisable()
    self:UnhookAll()
    self:UnregisterEvent('ACHIEVEMENT_SEARCH_UPDATED')
    self:UnregisterEvent('ACHIEVEMENT_EARNED')
    self.container:Hide()
    self.checkButtonAchievement:Hide()
end

function AheadOfTheCurve:ACHIEVEMENT_SEARCH_UPDATED()
    local numFiltered = GetNumFilteredAchievements();

    if numFiltered < 500 then
        for index in pairs(self.achievementSearchList) do
            self.achievementSearchList[index] = nil
        end
        
        for index = 1, numFiltered do
            local achievementId = GetFilteredAchievementID(index)
            local _, name, _, completed = GetAchievementInfo(achievementId)
            
            if completed then
                self.achievementSearchList[achievementId] = name
            end
        end
    end
    
    AceConfigRegistry:NotifyChange('AheadOfTheCurve')
end

function AheadOfTheCurve:ACHIEVEMENT_EARNED()
    self:GetHighestDefaultAchievement()
    AceConfigRegistry:NotifyChange('AheadOfTheCurve')
end

function AheadOfTheCurve:GetHighestDefaultAchievement()
    for index, instance in pairs(self.instances) do
        for _, id in pairs(instance.achievements) do
            local _, _, _, completed = GetAchievementInfo(id)
            
            if completed then
                self.instances[index].highestCompleted = id
                break
            end
        end
    end
end

function AheadOfTheCurve:DisplayHighestDefaultAchievement(index)
    if self.instances[index].highestCompleted then
        local _, name, _, _, _, _, _, _, _, iconPath = GetAchievementInfo(self.instances[index].highestCompleted)
        return name, iconPath
    else
        return nil, nil
    end
end

function AheadOfTheCurve:GetLFGCategory(findAGroupButton)
    local category = findAGroupButton:GetParent().selectedCategory;
    
    if category ~= 2 or category ~= 3 then
        self.container:Hide()
        self.checkButtonAchievement:Hide()
    else
        self.container:Show()
        self.checkButtonAchievement:Show()
    end

    if category == 2 then
        self:ScanBagsForKeystone()
    end
end

function AheadOfTheCurve:GetLFGInstance(signUpButton)
    local searchResultInfo = C_LFGList.GetSearchResultInfo(signUpButton:GetParent().selectedResult)
    local highestCompleted = self:GetLFGAchievement(searchResultInfo.activityID)
    local isMythicDungeon = self:IsMythicDungeon(searchResultInfo.activityID)
    local checkButtonAchievementText, checkButtonAchievementWidth, checkButtonAchievementPoint = self:GetCheckButtonAchievementData(isMythicDungeon)

    if not highestCompleted and not self.db.global.enable.override or not self.db.global.enable.override then
        self.container:Hide()
        self.checkButtonAchievement:Hide()
    else
        self.container:Show()
        self.checkButtonAchievement:Show()
        self.checkButtonAchievement:SetChecked(self.db.global.enable.whispers.achievement)
        self.checkButtonAchievement.text:SetText(checkButtonAchievementText)
        self.checkButtonAchievement.text:SetWidth(checkButtonAchievementWidth)
        self.checkButtonAchievement:SetPoint('CENTER', checkButtonAchievementPoint, 0)      
    end

    if isMythicDungeon and keyStoneLink then
        self:ShowMythicPlusOption(highestCompleted)
    else
        self.container:SetPoint('BOTTOM', 0, -55)
        self.container:SetSize(306, 50)
        self.checkButtonAchievement:SetPoint('CENTER', checkButtonAchievementPoint, 0)
        self.checkButtonKeystone:Hide()
        self.checkButtonKeystone:SetChecked(false)
    end
end

function AheadOfTheCurve:GetCheckButtonAchievementData(isMythicDungeon)
    if self.db.global.enable.override then
        return 'Send Override Achievement', 158, -85
    end

    if isMythicDungeon then
        return 'Send Mythic Plus Achievement', 170, -85
    end

    return 'Send Ahead of the Curve Achievement', 205, -100
end

function AheadOfTheCurve:GetLFGAchievement(instanceId)
    for _, instance in pairs(self.instances) do
        for _, id in pairs(instance.ids) do
            if id == instanceId then
                return instance.highestCompleted
            end
        end
    end
end

function AheadOfTheCurve:SendWhisper(signUpButton)
    local searchResultInfo = C_LFGList.GetSearchResultInfo(signUpButton:GetParent().resultID)

    if self.checkButtonAchievement:GetChecked() or self.checkButtonKeystone:GetChecked() then
        local achievementId
        
        if self.db.global.enable.override and self.db.global.overrideAchievement then
            achievementId = self.db.global.overrideAchievement
        else
            achievementId = self:GetLFGAchievement(searchResultInfo.activityID)
        end
        
        if achievementId then
            local success = pcall(SendChatMessage, self:BuildWhisper(achievementId), 'WHISPER', nil, searchResultInfo.leaderName)
            
            if not success then
                self:Print('There was an error sending your whisper.')
            end
        end
    end
end

function AheadOfTheCurve:BuildWhisper(achievementId)
    if self.checkButtonAchievement:GetChecked() and (self.checkButtonKeystone:GetChecked() and keyStoneLink) then
        return GetAchievementLink(achievementId) .. keyStoneLink
    end

    if self.checkButtonAchievement:GetChecked() then
        return GetAchievementLink(achievementId)
    end

    if self.checkButtonKeystone:GetChecked() and keyStoneLink then
        return keyStoneLink
    end
end

function AheadOfTheCurve:ScanBagsForKeystone()
    for bag = 0, 4 do
        local numSlots = GetContainerNumSlots(bag)

        for slot = 1, numSlots do
            local _, _, _, _, _, _, link, _, _, itemId = GetContainerItemInfo(bag, slot)

            if itemId == 138019 then
                keyStoneLink = link
                break
            end
        end
    end
end

function AheadOfTheCurve:IsMythicDungeon(instanceId)
    for _, id in pairs(self.instances[1].ids) do
        if id == instanceId then
            return true
        end
    end
end

function AheadOfTheCurve:ShowMythicPlusOption(highestCompleted)
     if not highestCompleted and not self.db.global.enable.override then
        self.checkButtonAchievement:Hide()
        self.container:SetPoint('BOTTOM', 0, -55)
        self.container:SetSize(306, 50)
        self.checkButtonKeystone:SetPoint('CENTER', -75, 0)
        self.container:Show()
    else
        self.container:SetPoint('BOTTOM', 0, -75)
        self.container:SetSize(306, 70)
        self.checkButtonAchievement:SetPoint('CENTER', -85, 9)
    end

    self.checkButtonKeystone:SetChecked(self.db.global.enable.whispers.keystone)    
    self.checkButtonKeystone:Show()
end


Modules/Options.lua
local AheadOfTheCurve = LibStub('AceAddon-3.0'):GetAddon('AheadOfTheCurve')
local AheadOfTheCurveOptions = AheadOfTheCurve:NewModule('AheadOfTheCurveOptions', 'AceConsole-3.0')
local AceConfig = LibStub('AceConfig-3.0')
local AceConfigDialog = LibStub('AceConfigDialog-3.0')
local AceConfigRegistry = LibStub('AceConfigRegistry-3.0')
local LDB = LibStub('LibDataBroker-1.1')
local icon = LibStub('LibDBIcon-1.0')
local newSearch = false
local optionsOpened = false

function AheadOfTheCurveOptions:OnInitialize()
    self.db = AheadOfTheCurve.db
    self:RegisterChatCommand('aotc', 'OpenOptions')

    local aotcLDB = LDB:NewDataObject('AheadOfTheCurveMinimap', {
        type = 'data source',
        text = 'AheadOfTheCurveMinimap',
        icon = 'Interface\\AddOns\\AheadOfTheCurve\\Media\\icon',
        OnClick = function(_, button) 
            if button == 'LeftButton' then
                self:OpenOptions() 
            end 
        end,
        OnTooltipShow = function(self)
            self:AddLine('Ahead of the Curve')
            self:AddLine('Left Click to open the options menu.')
        end
    })

    icon:Register('AheadOfTheCurveMinimap', aotcLDB, self.db.global.minimap)
end

function AheadOfTheCurveOptions:OnEnable()
    local options = self:GetOptions()
    local blizOptions = self:GetBlizOptions()

    self:GetInstanceOptions(options)
    
    AceConfig:RegisterOptionsTable('AheadOfTheCurve', options)
    AceConfig:RegisterOptionsTable('AheadOfTheCurveOptions', blizOptions)
    AceConfigDialog:AddToBlizOptions('AheadOfTheCurveOptions', 'Ahead of the Curve')
end

function AheadOfTheCurveOptions:TableLength(t)
    local count = 0

    for _ in pairs(t) do 
        count = count + 1 
    end

    return count
end

function AheadOfTheCurveOptions:LinkAchievement(index)
    local achievementId = AheadOfTheCurve.instances[index].highestCompleted
                        
    if achievementId then
        ChatFrame1EditBox:Show()
        ChatFrame1EditBox:SetFocus()
        ChatFrame1EditBox:Insert(GetAchievementLink(achievementId))
    end
end

function AheadOfTheCurveOptions:OpenOptions()
    if not optionsOpened then
        optionsOpened = true

        local height = 480 

        for index, instance in pairs(AheadOfTheCurve.instances) do
            if instance.highestCompleted ~= nil then
                height = height + 30
            end
        end

        AceConfigDialog:SetDefaultSize('AheadOfTheCurve', 500, height)
        AceConfigDialog:Open('AheadOfTheCurve')
    else
        optionsOpened = false
        AceConfigDialog:Close('AheadOfTheCurve')
    end
end

function AheadOfTheCurveOptions:GetBlizOptions()
    return {
        name = 'Ahead of the Curve',
        handler = AheadOfTheCurve,
        type = 'group',
        args = {
            configure = {
                name = 'Configure',
                descStyle = 'inline',
                type = 'execute',
                func = function()
                    while CloseWindows() do end
                    AheadOfTheCurveOptions:OpenOptions()
                end
            }
        }
    }
end

function AheadOfTheCurveOptions:GetOptions()
    return {
        name = 'Ahead of the Curve',
        handler = AheadOfTheCurve,
        type = 'group',
        args = {
            enable = {
                order = 0,
                name = 'Enable Addon',
                desc = 'Enables / disables the addon',
                type = 'toggle',
                get = function() return self.db.global.enable.addon end,
                set = function(info, value)
                    self.db.global.enable.addon = value
                    
                    if self.db.global.enable.addon then
                        AheadOfTheCurve:OnEnable()
                    else
                        AheadOfTheCurve:OnDisable()
                    end
                end
            },
            header1 = {
                order = 1,
                name = 'Default Highest Achievements Found',
                type = 'header'
            },
            header2 = {
                order = 2,
                name = 'Override Defaults',
                type = 'header'
            },
            enableOverride = {
                order = 2.1,
                name = 'Enable Override',
                desc = 'Overrides the default achievements found and will always send the selected achievement from the dropdown.',
                type = 'toggle',
                disabled = function() return not self.db.global.enable.addon end,
                get = function() return self.db.global.enable.override end,
                set = function(info, value)self.db.global.enable.override = value end
            },
            serach = {
                order = 2.2,
                name = 'Search Achievements',
                desc = 'Search term must be greater than 3 characters.',
                type = 'input',
                width = 'full',
                disabled = function() return not self.db.global.enable.override or not self.db.global.enable.addon end,
                set = function(info, value) 
                    SetAchievementSearchString(value) 
                    newSearch = true 
                end,
                validate = function(info, value) 
                    if string.len(value) < 3 then 
                        return 'Error: Search term must be greater than 3 characters' 
                    else 
                        return true 
                    end 
                end
            },
            results = {
                order = 2.3,
                name = function()
                    if newSearch then 
                        return string.format('Select Override Achievement: %s Results Returned', tostring(self:TableLength(AheadOfTheCurve.achievementSearchList)))
                    else
                        return 'Select Override Achievement'
                    end
                end,
                desc = 'Results are limited to 500 and only completed achievemnts. Please try a more specific search term if you cannot find the achievement listed.',
                type = 'select',
                values = AheadOfTheCurve.achievementSearchList,
                width = 'full',
                disabled = function() 
                    return not self.db.global.enable.override 
                           or not self.db.global.enable.addon 
                           or (not self.db.global.overrideAchievement and not newSearch) 
                end,
                get = function() 
                    if self.db.global.overrideAchievement then
                        return self.db.global.overrideAchievement
                    else
                        return 1
                    end
                end,
                set = function(info, value) self.db.global.overrideAchievement = value end,
                validate = function(info, value) 
                    if value == 1 then 
                        return 'Error: Please select an achievement' 
                    else 
                        return true 
                    end 
                end
            },
            header3 = {
                order = 3,
                name = 'Other Options',
                type = 'header'
            },
            enableWhisperAchievement = {
                order = 3.1,
                name = 'Always Check Achievement Whisper Dialog Checkbox',
                desc = 'This will always check the achievement whisper dialog checkbox when signing up for a group by default.',
                type = 'toggle',
                width = 'double',
                disabled = function() return not self.db.global.enable.addon end,
                get = function() return self.db.global.enable.whispers.achievement end,
                set = function(info, value) 
                    self.db.global.enable.whispers.achievement = value 
                    AheadOfTheCurve.checkButtonAchievement:SetChecked(value) 
                end,
            },
            enableWhisperKeystone = {
                order = 3.2,
                name = 'Always Check Keystone Whisper Dialog Checkbox',
                desc = 'This will always check the keystone whisper dialog checkbox when signing up for a mythic plus group by default.',
                type = 'toggle',
                width = 'double',
                disabled = function() return not self.db.global.enable.addon end,
                get = function() return self.db.global.enable.whispers.keystone end,
                set = function(info, value) 
                    self.db.global.enable.whispers.keystone = value 
                    AheadOfTheCurve.checkButtonKeystone:SetChecked(value)  
                end,
            },
            minimap = {
                order = 3.3,
                name = 'Show Minimap Icon',
                desc = 'Displays the minimap icon.',
                type = 'toggle',
                disabled = function() return not self.db.global.enable.addon end,
                get = function() return not self.db.global.minimap.hide end,
                set = function(info, value) 
                    self.db.global.minimap.hide = not value 

                    if value then 
                        icon:Show('AheadOfTheCurveMinimap') 
                    else 
                        icon:Hide('AheadOfTheCurveMinimap') 
                    end 
                end
            },
            header4 = {
                order = 4,
                name = 'About',
                type = 'header'
            },
            about = {
                order = 4.1,
                name = 'Version: @project-version@\nCreated by Pigletoos of Zul\'jin-US',
                type = 'description'
            }
        }
    }
end

function AheadOfTheCurveOptions:GetInstanceOptions(options)
    local highestCompletedCount = 0

    for index, instance in pairs(AheadOfTheCurve.instances) do
        if next(instance.achievements) ~= nil and instance.highestCompleted ~= nil then
            options.args[string.format('instance%sIcon', index)] = {
                order = 1 + (index / 10),
                image = function() local _, iconPath = AheadOfTheCurve:DisplayHighestDefaultAchievement(index) return iconPath end,
                type = 'execute',
                width = 'half',
                name = '',
                func = function() self:LinkAchievement(index) end
            }

            options.args[string.format('instance%sButton', index)] = {
                order = 1 + ((index / 10) + 0.01),
                name = function() local name = AheadOfTheCurve:DisplayHighestDefaultAchievement(index) return string.format('%s: %s', instance.name, name) end,
                desc = 'Link Achievement',
                type = 'execute',
                width = 'double',
                func = function() self:LinkAchievement(index) end
            }

            highestCompletedCount = highestCompletedCount + 1
        end
    end

    if highestCompletedCount == 0 then
        local instances = {}

        for index, instance in pairs(AheadOfTheCurve.instances) do
            if next(instance.achievements) ~= nil then
                table.insert(instances, string.format('\n - %s', instance.name))
            end
        end

        options.args.noneFound = {
            order = 1.1,
            name = string.format('No achievements have been earned for: %s', table.concat(instances)),
            width = 'full',
            type = 'description'
        }
    end
end
